File: solutions/022.Generate_Parentheses/AC_dfs_less2^n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_less2^n.java
*  Create Date: 2015-03-05 10:20:26
*  Descripton:   
*/

import java.util.*;

public class Solution {

    public List<String> generateParenthesis(int n) {
        List<String> ret = new ArrayList<String>(), inner, outter;
        if (n == 0) {
            ret.add("");
            return ret;
        }
        if (n == 1) {
            ret.add("()");
            return ret;
        }
        for (int i = 0; i < n; ++i) {
            inner = generateParenthesis(i);
            outter = generateParenthesis(n - i - 1);
            for (int j = 0; j < inner.size(); ++j) {
                for (int k = 0; k < outter.size(); ++k) {
                    ret.add("(" + inner.get(j) + ")" + outter.get(k));
                }
            }
        }
        return ret;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 3, 4};
        System.out.println(s.generateParenthesis(3));
    }
}

File: solutions/022.Generate_Parentheses/AC_dfs_less2^n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_less2^n.py
# Create Date: 2015-03-05 10:27:16
# Usage:       AC_dfs_less2^n.py 
# Descripton:  


class Solution:
    # @param an integer
    # @return a list of string
    def generateParenthesis(self, n):
        dp = {0: [""], 1: ["()"]}

        def memorial_dfs(n):
            if n not in dp:
                dp[n] = []
                for i in range(n):
                    for inner in memorial_dfs(i):
                        for outter in memorial_dfs(n - i - 1):
                            dp[n].append('(' + inner + ')' + outter)
            return dp[n]

        return memorial_dfs(n)

# debug
s = Solution()
print s.generateParenthesis(3)
